File: 0835-image-overlap/0835-image-overlap.py
class Solution:
  def largestOverlap(self, img1: List[List[int]], img2: List[List[int]]) -> int:
    d = collections.defaultdict(int)
    ones1 = []
    ones2 = []
    n = len(img1)
    for i in range(n):
      for j in range(n):
        if img1[i][j] == 1:
          ones1.append((i, j))
        if img2[i][j] == 1:
          ones2.append((i, j))
    ans = 0
    for t1 in ones1:
      for t2 in ones2:
        t3 = (t2[0]-t1[0],t2[1]-t1[1])
        d[t3] += 1
        ans = max(ans, d[t3])
    return ans

File: 0835-image-overlap/0835-image-overlap.cpp
class Solution {
public:
  int largestOverlap(vector<vector<int>>& img1, vector<vector<int>>& img2) {
    int ans = 0;
    const int n = img1.size();
    vector<pair<int, int>> a, b;
    for (int i = 0; i < n; ++i) {
      for (int j = 0; j < n; ++j) {
        pair coord = make_pair(i, j);
        if (img1[i][j] == 1) a.push_back(coord);
        if (img2[i][j] == 1) b.push_back(coord);
      }
    }
    unordered_map<int, int> mp;
    for (const auto& aa : a) {
      for (const auto& bb : b) {
        ans = max(ans, ++mp[100*(aa.first-bb.first)+aa.second-bb.second]);
      }
    }
    return ans;
  }
};